iceoryx_hoofs/include/iceoryx_hoofs/internal/posix_wrapper/command_line_parser.hpp

```diff
@@ -0,0 +1,235 @@
+    ///        not provided
+    /// @param[in] argc number of arguments, see int main(int argc, char*argv[])
+    /// @param[in] argv the string array of arguments, see int main(int argc, char*argv[])
+    /// @param[in] argcOffset the starting point for the parsing. 1U starts at the first argument.
+    /// @param[in] actionWhenOptionUnknown defines the action which should be performed when the user sets a
+    ///             option/switch which is unknown
+    CommandLineOptions parse(int argc,
+                             char* argv[],
+                             const uint64_t argcOffset = 1U,
+                             const UnknownOption actionWhenOptionUnknown = UnknownOption::TERMINATE) noexcept;
+
+    struct Entry
+    {
+        char shortOption = NO_SHORT_OPTION;
+        CommandLineOptions::name_t longOption;
+        description_t description;
+        ArgumentType type = ArgumentType::SWITCH;
+        typeName_t typeName;
+        CommandLineOptions::value_t defaultValue;
+    };
+
+  private:
+    friend void internal::handleError(const CommandLineParser&);
+
+    CommandLineParser& addOption(const Entry& option) noexcept;
+    cxx::optional<Entry> getOption(const CommandLineOptions::name_t& name) const noexcept;
+    void printHelpAndExit() const noexcept;
+
+    /// BEGIN only used in parse to improve readability
+    ///
+    /// Do not use those internal methods outside of parse. They were written
+    /// to improve the readability of the code. None of those functions verify
+    /// the pre conditions they require, this has to be done by calling them
+    /// in the correct order.
+    bool areAllRequiredValuesPresent() const noexcept;
+    bool hasArguments(const int argc) const noexcept;
+    bool assignBinaryName(const char* name) noexcept;
+    bool doesOptionStartWithMinus(const char* option) const noexcept;
+    bool hasOptionName(const char* option) const noexcept;
+    bool hasValidSwitchName(const char* option) const noexcept;
+    bool hasValidOptionName(const char* option) const noexcept;
+    bool doesOptionNameFitIntoString(const char* option) const noexcept;
+    bool isNextArgumentAValue(const uint64_t position) const noexcept;
+    bool isOptionSet(const Entry& entry) const noexcept;
+    bool doesOptionValueFitIntoString(const char* value) const noexcept;
+    bool doesOptionHasSucceedingValue(const Entry& entry, const uint64_t position) const noexcept;
+    /// END only used in parse to improve readability
+
+    void sortAvailableOptions() noexcept;
+    void setDefaultValuesToUnsetOptions() noexcept;
+
+  private:
+    int m_argc = 0;
+    char** m_argv = nullptr;
+    uint64_t m_argcOffset = 0;
+
+    description_t m_programDescription;
+    cxx::vector<Entry, CommandLineOptions::MAX_NUMBER_OF_ARGUMENTS> m_availableOptions;
+    cxx::function<void()> m_onFailureCallback;
+    CommandLineOptions m_options;
```

Comment:
I somehow have problems with the terminology. For me, the options are the set of knobs I as a developer make accessible. The arguments are the knobs the user touches. Therefore it feels a bit weird to me to add switches and values to a parser and then get some options. For me it would make more sense to configure the `CommandLineOptions` with `addSwitch`, `addRequiredValue`, etc. and then pass this together with `argc` and `argv` to the `CommandLineParser`. The `CommandLineParser` has a `parse` method to return a `CommandLineArguments` struct which has methods like `get`. This would also nicely encapsulate the parser and prevent using methods which should only be used in `parse`, since there would only be this one method.

Additional fun fact, the current `CommandLineOptions` talks about access to the command line argument values in its documentation ;)

What do you think?

Reply:
I am struggling a bit with this. Yes it would be a cleaner design but you as a developer would never use it in this fashion since this is internal API. The only public interface is the struct one builds and which takes `argc` and `argv` as constructor arguments.

Reply:
I implemented your suggestion. The user creates a `CommandLineOptionSet`, this is used in `parseCommandLineArguments` as parameter to acquire a `CommandLineValue`.